textx/model.py

```diff
@@ -598,92 +599,129 @@ def call_obj_processors(metamodel, model_obj,
+                        for _, _, delayed \
+                                in m._tx_reference_resolver.delayed_crossrefs:
+                            line, col = parser.pos_to_linecol(delayed.position)
+                            error_text += ' "{}" of class "{}" at {}'.format(
+                                delayed.obj_name, delayed.cls.__name__, (
+                                    line, col))
+                    raise TextXSemanticError(error_text, line=line, col=col)
+
+                for m in models:
+                    # TODO: what does this check?
+                    assert not m._tx_reference_resolver.parser._inst_stack
+
+                # cleanup
+                for m in models:
+                    del m._tx_reference_resolver
+
+                # final check that everything went ok
+                for m in models:
+                    assert 0 == len(get_children_of_type(
+                        Postponed.__class__, m))
+
+                    # We have model loaded and all link resolved
+                    # So we shall do a depth-first call of object
+                    # processors if any processor is defined.
+                    if m._tx_metamodel.obj_processors:
+                        if parser.debug:
+                            parser.dprint("CALLING OBJECT PROCESSORS")
+                        call_obj_processors(m._tx_metamodel, m)
+
+            except BaseException as e:
+                # remove all processed models from (global) repo (if present)
+                # (remove all of them, not only the model with errors,
+                # since, models with errors may be included in other models)
+                models_to_be_removed = models
+                for m in models:
+                    remove_models_from_repositories(m, models_to_be_removed)
+                raise e
+
+        if metamodel.textx_tools_support \
+                and type(model) not in PRIMITIVE_PYTHON_TYPES:
+            # Cross-references for go-to definition language server support
+            # Already sorted based on ref_pos_start attr
+            # (required for binary search)
+            model._pos_crossref_list = pos_crossref_list
+
+            # Dict for storing rules where key is position of rule instance in
+            # text. Sorted based on nested rules.
+            model._pos_rule_dict = OrderedDict(sorted(pos_rule_dict.items(),
+                                                      key=lambda x: x[0],
+                                                      reverse=True))
+    # exception occurred during model creation
+    except BaseException as e:
+        # remove all models beeing constructed
+        all_affected_models = get_included_models(model)
+        models_to_be_removed = list(filter(
+                lambda x: hasattr(x, "_tx_reference_resolver"),
+                all_affected_models))
+        for m in all_affected_models:
+            remove_models_from_repositories(m, models_to_be_removed)
```

Comment:
Very good point. In order to increase the readability of the code I introduced `_start_model_construction` and `_remove_all_affected_models_in_construction` as suggested.

Because the decision of using the reference_resolver as "flag" for model to mark them as "being constructed" I left the two function in model.py as private functions. These can later easily be refactored...

I you prefer them somewhere else, I can move them.

Reply:
I also added `_end_model_construction`...